File: src/sag2grd/sag2grd.pl
#
# sag2grd
#
# convert simple ascii grid (sag) files to a pair of
# GMT netCDF (grd) files - this is done by piping the
# data to the GMT program xyz2grd 
#
# J.J. Green 2008
# $Id: sag2grd.pl,v 1.4 2010/08/09 19:20:39 jjg Exp jjg $

use strict;
use POSIX;
use Getopt::Std;

my $ist;
my @gsts = ();

my $usage = <<EOF
usage: sag2grd [options] <grd u> ...

where <grd u> ...  are the names of the grd files to be 
written and options can include

-h         : brief help
-i <file>  : read sag input from <file> rather than stdin
-v         : verbose
-V         : version
EOF
  ;

my %opt = ();
getopts('hi:vV',\%opt);

if ($opt{h})
{
    print $usage;
    exit EXIT_SUCCESS;
}

if ($opt{V})
{
    print "sag2grd $version\n";
    exit EXIT_SUCCESS;
}

sub info
{
    if ($opt{v})
    {
	print @_;
    }
}

info "This is sag2grd\n";

my $input = $opt{i} || "-";

unless (open $ist,"< $input")
{
    warn "failed to open $input to read\n";
    exit EXIT_FAILURE;
}

my $grdargs;
my $ngrd = scalar @ARGV;

if (defined(my $line = <$ist>))
{
    chomp $line;

    my ($magic,$ver,$gdim,$vdim,$gxn,$gyn,
	$gxmin,$gxmax,$gymin,$gymax) = split(/\s+/,$line);

    unless ($magic eq '#sag')
    {
	warn "bad magic $magic - not a sag file\n";
	exit EXIT_FAILURE;
    }

    if ($ver != 1)
    {
	warn "can't read sag version $ver\n";
	exit EXIT_FAILURE;
    }

    if ($gdim != 2)
    {
	warn "can't handle sag file with grid dimension $gdim\n";
	exit EXIT_FAILURE;
    }

    if ($ngrd != $vdim)
    {
	warn "sag file has vector dimension $vdim, but $ngrd grd files specified\n";
	exit EXIT_FAILURE;
    }

    $grdargs = "-R$gxmin/$gxmax/$gymin/$gymax -I$gxn+/$gyn+";
}

my @grds = @ARGV;

info "output to\n";

foreach my $grd (@grds)
{
    info "  $grd\n";
}

my $xyz2grd;

if ($xyz2grd = `which xyz2grd`)
{
    chomp $xyz2grd;
}
else
{
    my $wrapper = "GMT";

    if (my $gmtcmd = `which $wrapper`)
    {
	chomp $gmtcmd;
	$xyz2grd = "$gmtcmd xyz2grd";
    }
}

unless ($xyz2grd)
{
    warn "failed to find xyz2grd program";
    exit EXIT_FAILURE;
}

info "gridder is $xyz2grd $grdargs\n";

# create an array of stream to gridding processes,

foreach my $grd (@grds)
{
    my $gst;
    my $cmd = "| $xyz2grd -G$grd $grdargs";

    unless (open($gst,$cmd))
    {
	warn "failed to open process $cmd\n";
	exit EXIT_FAILURE;
    }

    push @gsts,$gst;
}

my $nline = 0;

sub trim
{
    my $s = shift;

    $s =~ s/^\s+//;
    $s =~ s/\s+$//;

    return $s;
}

while (defined(my $line = <$ist>))
{
    my ($x,$y,@u) = split(/\s+/,trim($line));

    for (my $i=0 ; $i<$ngrd ; $i++)
    {
	my $gst = $gsts[$i];
	printf $gst "%f %f %f\n",$x,$y,$u[$i];
    }

    $nline++;
}

info "wrote $nline points\n";

sub END
{
    my $err = 0;

    foreach my $gst (@gsts)
    { 
	close $gst;
    }

    if (defined $ist)
    {
	close $ist;
    }

    info "done.\n";
}

File: src/sag2grd/sag2grd.py
#!/usr/bin/env python3

import argparse as ap
import sys
import shutil
import subprocess as sp


def sag2grd_st(args, st):
    line = st.readline().rstrip()
    parts = line.split()
    if len(parts) != 11:
        raise RuntimeError('bad SAG header "%s"' % line)
    magic, ver, gdim, vdim, gxn, gyn, gxmin, gxmax, gymin, gymax, tol = parts

    if magic != '#sag':
        raise RuntimeError('bad magic "%s", not a SAG file' % magic)

    ver = int(ver)
    if ver != 1:
        raise RuntimeError('version %i not supported' % ver)

    gdim = int(gdim)
    if gdim != 2:
        raise RuntimeError('grid dimension %i not supported' % gdim)

    ngrd = len(args.grd)
    vdim = int(vdim)
    if ngrd != vdim:
        raise RuntimeError('SAG has vector dimension %i but %i grd files given'
                           % (vdim, ngrd))

    grdargs = ['-R%s/%s/%s/%s' % (gxmin, gxmax, gymin, gymax),
               '-I%s+/%s+' % (gxn, gyn)]

    info('output to:')
    for grd in args.grd:
        info('  %s' % grd)

    xyz2grd = shutil.which('xyz2grd')
    if xyz2grd:
        cmd = [xyz2grd]
    else:
        wrapper = shutil.which('GMT')
        if not wrapper:
            raise RuntimeError('neither "xyz2grd" nor "GMT" found')
        cmd = [wrapper, 'xyz2grd']

    info('gridder is %s' % ' '.join(cmd + grdargs))

    gridders = []
    for grd in args.grd:
        gridder = sp.Popen(cmd + ['-G%s' % grd] + grdargs, stdin=sp.PIPE)
        gridders.append(gridder)

    count = 0
    for line in st:
        parts = line.rstrip().split()
        x, y, *vals = parts
        for i in range(ngrd):
            data = bytes(' '.join([x, y, vals[i]]) + "\n", 'ascii')
            gridders[i].stdin.write(data)
        count += 1

    info("wrote %i points" % count)

    for gridder in gridders:
        gridder.stdin.close()
        gridder.wait()

    return 0


def sag2grd(args):
    info('This is sag2grd')

    if len(args.input) > 0:
        info('reading from %s' % (args.input[0]))
        st = open(args.input[0], 'r')
        err = sag2grd_st(args, st)
        st.close()
    else:
        err = sag2grd_st(args, sys.stdin)

    return err

parser = ap.ArgumentParser(prog='sag2grd')
parser.add_argument('--input', '-i', metavar='<sag>', nargs=1,
                    help='read sag input from file')
parser.add_argument('--verbose', '-v', action='store_true',
                    help='verbose operation')
parser.add_argument('--version', '-V', action='version', version='1.0.12')
parser.add_argument('grd', nargs='+', metavar='<grd>', help='GMT grd file')

args = parser.parse_args()

if args.verbose:
    def info(a):
        print(a)
else:
    def info(a):
        pass

err = sag2grd(args)

if err:
    info('failed.')
else:
    info('done.')

exit(err)
